joint_state_controller:
  type: joint_state_controller/JointStateController
  publish_rate: &loop_hz 500

pos_joint_traj_controller:
  type: position_controllers/JointTrajectoryController
  joints: &robot_joints
    - waist
    - shoulder
    - elbow
    - forearm_roll
    - wrist_angle
    - wrist_rotate  
  constraints:
    goal_time: 0.6
    stopped_velocity_tolerance: 0.05
    waist: { trajectory: 0.2, goal: 0.1 }
    shoulder: { trajectory: 0.2, goal: 0.1 }
    elbow: { trajectory: 0.2, goal: 0.1 }
    forearm_roll: { trajectory: 0.2, goal: 0.1 }
    wrist_angle: { trajectory: 0.2, goal: 0.1 }
    wrist_rotate: { trajectory: 0.2, goal: 0.1 }
  stop_trajectory_duration: 0.5
  state_publish_rate: *loop_hz
  action_monitor_rate: 20

gripper_action_controller:
  type: position_controllers/GazeboGripperActionController
  joint: &gripper_joints
    - left_finger
  goal_tolerance: 0.1
  max_effort: 100.0
  stall_position_threshold: 0.05 # Tune to fit your robot
  stall_timeout: 0.5
  action_monitor_rate: 10
  
joint_group_pos_controller:
  type: position_controllers/JointGroupPositionController
  joints: *robot_joints
  pid: {p: 0.01, i: 0.0001, d: 0.001}

gripper_pos_controller:
  type: position_controllers/JointPositionController
  joint: *gripper_joints

effort_joint_group_pos_controller:
  type: effort_controllers/JointGroupPositionController
  joints: *robot_joints
  pid: {p: 1.0, i: 0.01, d: 0.1}

motion_control_handle:
  type: "cartesian_controllers/MotionControlHandle"
  end_effector_link: &ee_link "tcp_link"
  robot_base_link: &base_link "vx300s/base_link"
  target_frame_topic: "/motion_control_handle/target_frame"
  joints: *robot_joints

cartesian_motion_controller:
  type: "position_controllers/CartesianMotionController"
  end_effector_link: *ee_link
  robot_base_link: *base_link
  target_frame_topic: "target_frame"
  joints: *robot_joints

  pd_gains:
    trans_x: { p: 10.0 }
    trans_y: { p: 10.0 }
    trans_z: { p: 10.0 }
    rot_x: { p: 1.0 }
    rot_y: { p: 1.0 }
    rot_z: { p: 1.0 }

cartesian_force_controller:
  type: "position_controllers/CartesianForceController"
  end_effector_link: *ee_link
  robot_base_link: *base_link
  ft_sensor_ref_link: "wrist_rotate"
  joints: *robot_joints

  command_interfaces:
    #- position
    - velocity

  solver:
    error_scale: 1.0

  pd_gains:
    trans_x: { p: 0.1 }
    trans_y: { p: 0.1 }
    trans_z: { p: 0.1 }
    rot_x: { p: 1.0 }
    rot_y: { p: 1.0 }
    rot_z: { p: 1.0 }

cartesian_compliance_controller:
  type: "position_controllers/CartesianComplianceController"
  end_effector_link: *ee_link
  robot_base_link: *base_link
  ft_sensor_ref_link: "wrist_rotate"
  compliance_ref_link: "tcp_link"
  target_frame_topic: "target_frame"
  joints: *robot_joints

  stiffness: # w.r.t. compliance_ref_link
    trans_x: 100
    trans_y: 100
    trans_z: 100
    rot_x: 5
    rot_y: 5
    rot_z: 5

  solver:
    error_scale: 1.0

  pd_gains:
    trans_x: { p: 0.2 }
    trans_y: { p: 0.2 }
    trans_z: { p: 0.2 }
    rot_x: { p: 2.0 }
    rot_y: { p: 2.0 }
    rot_z: { p: 2.0 }

compliant_traj_action_server:
  end_effector_link: *ee_link
  robot_base_link: *base_link
  joints: *robot_joints
  path_tolerance:
    pose_error: {trans_x: 0.3, trans_y: 0.3, trans_z: 0.3, rot_x: 1.0, rot_y: 1.0, rot_z: 1.0}
    twist_error: {trans_x: 1000000, trans_y: 1000000, trans_z: 1000000, rot_x: 1000000, rot_y: 1000000, rot_z: 1000000}
    accel_error: {trans_x: 1000000, trans_y: 1000000, trans_z: 1000000, rot_x: 1000000, rot_y: 1000000, rot_z: 1000000}
  goal_tolerance:
    pose_error: {trans_x: 0.2, trans_y: 0.2, trans_z: 0.2, rot_x: 0.5, rot_y: 0.5, rot_z: 0.5}
    twist_error: {trans_x: 1000000, trans_y: 1000000, trans_z: 1000000, rot_x: 1000000, rot_y: 1000000, rot_z: 1000000}
    accel_error: {trans_x: 1000000, trans_y: 1000000, trans_z: 1000000, rot_x: 1000000, rot_y: 1000000, rot_z: 1000000}
  goal_time_tolerance: 3.0

gazebo_ros_control:
  pid_gains:
    # arm
    # waist: { p: 4000, d: 200, i: 1, i_clamp: 1 }
    # shoulder: { p: 10000, d: 200, i: 1, i_clamp: 1 }
    # elbow: { p: 3000, d: 20, i: 1, i_clamp: 1 }
    # forearm_roll: { p: 1000, d: 1, i: 1, i_clamp: 1 }
    # wrist_angle: { p: 500, d: 1, i: 1, i_clamp: 1 }
    # wrist_rotate: { p: 50, d: 0.1, i: 0, i_clamp: 1 }
    # gripper
    left_finger: {p: 10,  d: 0.1, i: 1, i_clamp: 1}
    right_finger: {p: 10,  d: 0.1, i: 1, i_clamp: 1}